File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/Executable.java
package org.hyperledger.iroha.android.model;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/** Represents the executable payload embedded in a transaction. */
public final class Executable {

  private enum Variant {
    INSTRUCTIONS,
    CONTRACT_CALL,
    BATCH,
    IVM
  }

  private final Variant variant;
  private final List<InstructionBox> instructions;
  private final ContractInvocation contractInvocation;
  private final List<ExecutableBatchItem> batchItems;
  private final byte[] ivmBytes;

  private Executable(
      final Variant variant,
      final List<InstructionBox> instructions,
      final ContractInvocation contractInvocation,
      final List<ExecutableBatchItem> batchItems,
      final byte[] ivmBytes) {
    this.variant = variant;
    this.instructions = immutableCopy(instructions, "instruction");
    this.contractInvocation = contractInvocation;
    this.batchItems = immutableCopy(batchItems, "batchItem");
    this.ivmBytes = ivmBytes == null ? new byte[0] : Arrays.copyOf(ivmBytes, ivmBytes.length);
  }

  public static Executable instructions(final List<? extends InstructionBox> instructions) {
    Objects.requireNonNull(instructions, "instructions");
    return new Executable(
        Variant.INSTRUCTIONS, new ArrayList<>(instructions), null, null, null);
  }

  public static Executable contractCall(final ContractInvocation invocation) {
    return new Executable(
        Variant.CONTRACT_CALL,
        null,
        Objects.requireNonNull(invocation, "contractInvocation"),
        null,
        null);
  }

  /** Creates an ordered, flat mix of instructions and deployed-contract calls. */
  public static Executable batch(final List<? extends ExecutableBatchItem> items) {
    Objects.requireNonNull(items, "items");
    if (items.isEmpty()) {
      throw new IllegalArgumentException("executable batch must contain at least one item");
    }
    return new Executable(Variant.BATCH, null, null, new ArrayList<>(items), null);
  }

  public static Executable ivm(final byte[] ivmBytes) {
    Objects.requireNonNull(ivmBytes, "ivmBytes");
    return new Executable(Variant.IVM, null, null, null, ivmBytes);
  }

  public boolean isInstructions() {
    return variant == Variant.INSTRUCTIONS;
  }

  public boolean isIvm() {
    return variant == Variant.IVM;
  }

  public boolean isContractCall() {
    return variant == Variant.CONTRACT_CALL;
  }

  public boolean isBatch() {
    return variant == Variant.BATCH;
  }

  /** Returns the instruction list. Typed payloads are hydrated when bindings are available. */
  public List<InstructionBox> instructions() {
    if (!isInstructions()) {
      return Collections.emptyList();
    }
    return instructions;
  }

  public ContractInvocation contractInvocation() {
    if (!isContractCall()) {
      throw new IllegalStateException("Executable does not contain a contract call");
    }
    return contractInvocation;
  }

  /** Returns the ordered mixed-batch items, or an empty list for another variant. */
  public List<ExecutableBatchItem> batchItems() {
    if (!isBatch()) {
      return Collections.emptyList();
    }
    return batchItems;
  }

  public byte[] ivmBytes() {
    if (!isIvm()) {
      throw new IllegalStateException("Executable does not contain IVM bytecode");
    }
    return Arrays.copyOf(ivmBytes, ivmBytes.length);
  }

  /** Whether this executable requires a signature-bound transaction gas limit. */
  public boolean requiresTransactionGasLimit() {
    if (isContractCall() || isIvm()) {
      return true;
    }
    if (isBatch()) {
      for (final ExecutableBatchItem item : batchItems) {
        if (item.isContractCall()) {
          return true;
        }
      }
    }
    return false;
  }

  @Override
  public boolean equals(final Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof Executable)) {
      return false;
    }
    final Executable that = (Executable) other;
    return variant == that.variant
        && instructions.equals(that.instructions)
        && Objects.equals(contractInvocation, that.contractInvocation)
        && batchItems.equals(that.batchItems)
        && Arrays.equals(ivmBytes, that.ivmBytes);
  }

  @Override
  public int hashCode() {
    int result = Objects.hash(variant, instructions, contractInvocation, batchItems);
    result = 31 * result + Arrays.hashCode(ivmBytes);
    return result;
  }

  private static <T> List<T> immutableCopy(final List<? extends T> values, final String field) {
    if (values == null) {
      return Collections.emptyList();
    }
    final ArrayList<T> copy = new ArrayList<>(values.size());
    for (final T value : values) {
      copy.add(Objects.requireNonNull(value, field));
    }
    return Collections.unmodifiableList(copy);
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/Executable.kt
package org.hyperledger.iroha.sdk.core.model

/**
 * Represents the executable payload embedded in a transaction.
 */
sealed class Executable {

    class Instructions(instructions: List<InstructionBox>) : Executable() {

        private val _instructions: List<InstructionBox> =
            java.util.Collections.unmodifiableList(ArrayList(instructions))

        val instructions: List<InstructionBox> get() = _instructions

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Instructions) return false
            return _instructions == other._instructions
        }

        override fun hashCode(): Int = _instructions.hashCode()
    }

    class Ivm(ivmBytes: ByteArray) : Executable() {

        private val _ivmBytes: ByteArray = ivmBytes.copyOf()

        val ivmBytes: ByteArray get() = _ivmBytes.copyOf()

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Ivm) return false
            return _ivmBytes.contentEquals(other._ivmBytes)
        }

        override fun hashCode(): Int = _ivmBytes.contentHashCode()
    }

    /** Invokes one deployed contract instance by reference. */
    class ContractCall(@JvmField val invocation: ContractInvocation) : Executable() {
        override fun equals(other: Any?): Boolean =
            this === other || other is ContractCall && invocation == other.invocation

        override fun hashCode(): Int = invocation.hashCode()
    }

    /** Ordered, atomic mix of native instructions and deployed-contract invocations. */
    class Batch(entries: List<ExecutableBatchItem>) : Executable() {
        init {
            require(entries.isNotEmpty()) { "executable batch must contain at least one item" }
        }

        private val _entries: List<ExecutableBatchItem> =
            java.util.Collections.unmodifiableList(ArrayList(entries))

        val entries: List<ExecutableBatchItem> get() = _entries

        override fun equals(other: Any?): Boolean =
            this === other || other is Batch && _entries == other._entries

        override fun hashCode(): Int = _entries.hashCode()
    }

    /** Mutable authoring helper that preserves the exact order in which batch items are added. */
    class BatchBuilder internal constructor() {
        private val entries: MutableList<ExecutableBatchItem> = mutableListOf()

        fun addInstruction(instruction: InstructionBox) = apply {
            entries.add(ExecutableBatchItem.instruction(instruction))
        }

        fun addContractCall(invocation: ContractInvocation) = apply {
            entries.add(ExecutableBatchItem.contractCall(invocation))
        }

        fun addItem(item: ExecutableBatchItem) = apply {
            entries.add(item)
        }

        fun build(): Executable = Batch(entries)
    }

    /** Whether this executable requires a signature-bound gas limit in its fee intent. */
    fun requiresTransactionGasLimit(): Boolean = when (this) {
        is Instructions -> false
        is Ivm, is ContractCall -> true
        is Batch -> entries.any { it is ExecutableBatchItem.ContractCall }
    }

    companion object {
        @JvmStatic
        fun instructions(instructions: List<InstructionBox>): Executable =
            Instructions(instructions)

        @JvmStatic
        fun ivm(ivmBytes: ByteArray): Executable = Ivm(ivmBytes)

        @JvmStatic
        fun contractCall(invocation: ContractInvocation): Executable = ContractCall(invocation)

        @JvmStatic
        fun batch(entries: List<ExecutableBatchItem>): Executable = Batch(entries)

        @JvmStatic
        fun batchBuilder(): BatchBuilder = BatchBuilder()
    }
}
